Add query filter icons to filterable table fields in explore section

miqh opened the pull request:
Hello,

This pull request attempts to address #13520.

Apart from introducing the query filter icons, the only other notable change was updating a function used to modify queries to apply filtering. If applying a filter concerning a label selector already present in a query, the incoming filter should replace the existing one (i.e. doesn't make sense to specify the same selector more than once, `{job="a", job="a"}`).

davkal commented on public/app/features/explore/Explore.tsx:
```diff
@@ -416,8 +416,13 @@ export class Explore extends React.PureComponent<ExploreProps, ExploreState> {
+  onClickTableCell = (columnKey: string, rowValue: string, attributes: {}) => {
+    this.onModifyQueries({
+      key: columnKey,
+      operator: attributes['data-operator'],
```
Explore should be agnostic about what it means to add a filter. Hence I would pass the operator on in an options map. Then it's for the datasource to decode what it needs from the options.

davkal commented on public/app/features/explore/Table.tsx:
```diff
@@ -28,7 +29,11 @@ export default class Table extends PureComponent<TableProps> {
             const columnKey = column.Header;
             const rowValue = rowInfo.row[columnKey];
-            this.props.onClickCell(columnKey, rowValue);
+            const attributes = Array.from(link.attributes).reduce((attributes, attribute) => {
```
I'm not a huge fan of data attributes, but ReactTable is already doing something weird here anyway by channelling all cell events through here, so I think it's actually the least painful way to get input to a handler.

davkal commented on public/app/features/explore/Table.tsx:
```diff
@@ -44,7 +49,25 @@ export default class Table extends PureComponent<TableProps> {
+          {row.value}
+          {filterable && (
+            <span className="filter-links">
+              <Tooltip content="Filter for value" placement="bottom">
```
The tooltip gets in the way for me by appearing and disappearing quickly, so I can't even click on the icon. I would definitely prefer a simple element title here for native tooltips.

miqh replied:
@davkal, I also noticed this as I was making changes, but couldn't reliably reproduce it. :thinking:

Had a quick look at the `Tooltip` implementation which eventually led to Popper.js, and I couldn't immediately see anything wrong. I'll replace with a simple `title` attribute for now. Perhaps the flicker is a bug worth a look later.

davkal commented on public/app/features/explore/Table.tsx:
```diff
@@ -44,7 +49,25 @@ export default class Table extends PureComponent<TableProps> {
+          {filterable && (
+            <span className="filter-links">
+              <Tooltip content="Filter for value" placement="bottom">
+                <a className="link" data-operator="=">
```
The Table should be a generic viewer of data and should not prescribe what operator to use for the filter. I'd rather pass a constant string here like `INCLUDE` and let the datasource interpret that when it needs to modify the query.

davkal commented on public/app/plugins/datasource/prometheus/add_label_to_query.ts:
```diff
@@ -66,7 +66,11 @@ function addLabelToSelector(selector: string, labelKey: string, labelValue: stri
-      parsedLabels.push({ key: match[1], operator: match[2], value: match[3] });
+      const key = match[1];
+      // Ignore existing label selector when the incoming one should replace it
+      if (key !== labelKey) {
```
Good point. Would be good to add a test for this.

davkal commented on public/app/plugins/datasource/prometheus/datasource.ts:
```diff
@@ -456,7 +456,7 @@ export class PrometheusDatasource {
     switch (action.type) {
       case 'ADD_FILTER': {
-        return addLabelToQuery(query, action.key, action.value);
+        return addLabelToQuery(query, action.key, action.value, action.operator);
```
Here I would interpret the action option to map it to an operator.

davkal commented on public/sass/pages/_explore.scss:
```diff
@@ -383,6 +383,25 @@
   top: 60px;
   z-index: inherit;
 }
+.ReactTable .filter-links {
```
Any chance we can reuse `table-panel-cell-filterable` and `table-panel-filter-link` styles here? They are defined by the dashboard table panel plugin.

miqh replied:
@davkal, yep, can do. I opted for the new styles alongside the existing `.ReactTable` ones just for cohesiveness with this component definition.

`.table-panel-filter-link` looks like a drop-in replacement, but some additional fidgeting is needed to get the hover and visibility going. That'd mean applying `.table-panel-table` on some parent. Do you intend for `.ReactTable` to eventually reuse everything from `_panel_table.scss`? Please correct me if I misunderstood here.

davkal replied:
We're still testing react-table and are not completely decided on it. It'd be better to make it work with the grafana styles.

davkal reviewed (changes requested):
Nice and straightforward approach. It's crucial though, to not tie Table/Graph and other result viewers to a specific datasource. 
Couldnt actually test the feature since the tooltip kept overlaying over the link.

miqh committed:
Add icons to filterable table fields in explore section

The icons behave similar to those in the dashboard table panel and
apply an inclusive or exclusive query selector filter.

Related: #13520
